src/main/scala/tilelink/Parameters.scala

```diff
@@ -318,6 +357,44 @@ case class TLClientPortParameters(
+  def emitsLogicalSafe   (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(true, _.knownToEmit.get.emitsLogical,    address, lgSize, range)
+  def emitsGetSafe       (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(true, _.knownToEmit.get.emitsGet,        address, lgSize, range)
+  def emitsPutFullSafe   (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(true, _.knownToEmit.get.emitsPutFull,    address, lgSize, range)
+  def emitsPutPartialSafe(address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(true, _.knownToEmit.get.emitsPutPartial, address, lgSize, range)
+  def emitsHintSafe      (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(true, _.knownToEmit.get.emitsHint,       address, lgSize, range)
+
+  def emitsAcquireTFast  (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsAcquireT,   address, lgSize, range)
+  def emitsAcquireBFast  (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsAcquireB,   address, lgSize, range)
+  def emitsArithmeticFast(address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsArithmetic, address, lgSize, range)
+  def emitsLogicalFast   (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsLogical,    address, lgSize, range)
+  def emitsGetFast       (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsGet,        address, lgSize, range)
+  def emitsPutFullFast   (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsPutFull,    address, lgSize, range)
+  def emitsPutPartialFast(address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsPutPartial, address, lgSize, range)
+  def emitsHintFast      (address: UInt, lgSize: UInt, range: Option[TransferSizes] = None) = if (clients.exists(_.knownToEmit == None)) true.B else emitHelper(false, _.knownToEmit.get.emitsHint,       address, lgSize, range)
+
```

Comment:
There's a lot of duplication here, you may notice the lines around this where the code does stuff like:
```scala
val supportsProbe      = safety_helper(_.supportsProbe)      _
```
That's using currying to duplicate less, and I would suggest doing the same thing here.

eg.

Assuming we can unify `emitHelper` with `supportHelper`, we can wrap with a curried function:

```scala
  private def curriedEmitHelper(
      safe:    Boolean,
      f:  TLManagerEmissionSizes => TransferSizes)
     (address: UInt,
      lgSize:  UInt,
      range:   Option[TransferSizes] = None): Bool =
  if (clients.exists(_.knownToEmit == None)) true.B
  else emitHelper(safe, p => f(p.knownToEmit.get), address, lgSize, range)
```

Then you should be able to vastly simplify these functions:
```scala
val emitsAcquireTSafe = curriedEmitHelper(true, _.emitsAcquireT) _
...
```

Reply:
Did you want me to change the supportsAcquireTSafe... functions as well?

Reply:
Ah I see you were copying that style, I'll defer to @terpstra on if those should be changed.

I think it's more important here because you have the additional `if (clients.exists(_.knownToEmit == None)) true.B else ...` boilerplate.